src/compiler/checker.ts

```diff
@@ -16089,6 +16089,21 @@ namespace ts {
             return false;
         }
 
+        function getJavaScriptClassType(symbol: Symbol): Type | undefined {
+            if (symbol && isDeclarationOfFunctionOrClassExpression(symbol)) {
+                symbol = getSymbolOfNode((<VariableDeclaration>symbol.valueDeclaration).initializer);
+            }
+            if (isJavaScriptConstructor(symbol.valueDeclaration)) {
+                return getInferredClassType(symbol);
+            }
+            if (symbol.flags & SymbolFlags.Variable) {

```

Comment:
Line 16096 handles the case where `symbol` refers directly (globally or within the same file) to a function or a variable declaration whose initializer is a function. 

Line 16099 handles the case where `symbol` refers indirectly to a function or variable declaration whose initializer is a function. This is usually the case for imports like `const C = require("./c")` or `const { C } = require("./c")`, but also cases like `const C = B` where `B` is also local.